fxtest/fxtest.go

```diff
@@ -0,0 +1,94 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package fxtest
+
+import (
+	"context"
+
+	"go.uber.org/fx"
+)
+
+type starter interface {
+	Start() error
+}
+
+type stopper interface {
+	Stop() error
+}
+
+// Fataler is a subset of the standard library's testing.TB interface. It's
+// satisfied by both *testing.T and *testing.B.
+type Fataler interface {
+	Fatalf(string, ...interface{})
+}
+
+// Lifecycle is a testing spy for fx.Lifecycle. It exposes Start and Stop
+// methods (and some test-specific helpers) so that unit tests can exercise
+// hooks.
+type Lifecycle struct {
+	fx.Lifecycle
+
+	t Fataler
+}
+
+// NewLifecycle creates a new test lifecycle.
+func NewLifecycle(t Fataler) *Lifecycle {
+	var deps struct {
+		fx.Lifecycle
+	}
+	app := fx.New(fx.Inject(&deps))
+	if err := app.Start(context.Background()); err != nil {
+		t.Fatalf("failed to create new lifecycle: %v", err)
+	}
+	if err := app.Stop(context.Background()); err != nil {
+		t.Fatalf("failed to create new lifecycle: %v", err)
```

Comment:
None of that is part of the public API, so I wouldn't call it a hard requirement. The current implementation is safe to start and stop repeatedly, so this seemed reasonable to me.

I don't mind moving the lifecycle implementation into a separate package, but it's going to move around a bunch more code.